target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c

```diff
@@ -2396,23 +2397,34 @@ static void rtl930x_led_init(struct rtl838x_switch_priv *priv)
 		return;
 	}
 
+	for (int i = 0; i < 4; i++) {
+		const __be32 *led_set;
+		char set_name[9];
+		u32 setlen;
+		u32 v;
+
+		sprintf(set_name, "led_set%d", i);
+		led_set = of_get_property(node, set_name, &setlen);
+		if (!led_set || setlen != 16)
+			break;
+		led_sets = i;
+		v = be32_to_cpup(led_set) << 16 | be32_to_cpup(led_set + 1);
+		sw_w32(v, RTL930X_LED_SET0_0_CTRL - 4 - i * 8);
+		v = be32_to_cpup(led_set + 2) << 16 | be32_to_cpup(led_set + 3);
+		sw_w32(v, RTL930X_LED_SET0_0_CTRL - i * 8);
+	}
+
 	for (int i = 0; i < priv->cpu_port; i++) {
 		int pos = (i << 1) % 32;
 		u32 set;
-		u32 v;
 
 		sw_w32_mask(0x3 << pos, 0, RTL930X_LED_PORT_FIB_SET_SEL_CTRL(i));
 		sw_w32_mask(0x3 << pos, 0, RTL930X_LED_PORT_COPR_SET_SEL_CTRL(i));
 
 		if (!priv->ports[i].phy)
 			continue;
 
-		v = 0x1;
-		if (priv->ports[i].is10G)
-			v = 0x3;
-		if (priv->ports[i].phy_is_integrated)
-			v = 0x1;
-		sw_w32_mask(0x3 << pos, v << pos, RTL930X_LED_PORT_NUM_CTRL(i));
+		sw_w32_mask(0x3 << pos, led_sets << pos, RTL930X_LED_PORT_NUM_CTRL(i));
```

Comment:
Okay I checked again, and I think the DTS, my code and old code all are wrong,
See, https://gitlab.com/olliver/openwrt/realtek_sdk/-/blob/xgs1250/sdk/system/include/hwp/hw_profiles/xgs1250_12.c?ref_type=heads#L36

There are 3 LED sets for xgs1250,
led_set0 -> has 2 LEDs and 0xffff because of `HWP_LED_END`
led_set1 -> has 4 LEDs
led_set2 -> has 2 LEDs and 0xffff because of `HWP_LED_END`

0xffff can also be read as enable LED to indicate everything

Looking at Datasheet of xgs1250-12, https://www.zyxelguard.com/datasheets/XGS1250-12_1.pdf
- for 1G ports, 2 leds per port 1-8. 
  - Link/Act + 10/100 M
  - 1000 M
- for 10G ports, 5 color LEDs per port 9,10,11 - which is achieved by combining 4 color LEDs 
  - 100M
  - 1000M
  - 2.5G
  - 5G
  - 10G
- for 10G sfp port
  - 1000M
  - 10G

I don't have xgs1250 to see how it works, i will rework LED code and test on a dev board that I have and back, after someone can compile and check on their xgs1250

Reply:
reworked led code, check latest push